File: contracts/jcs.py
"""
RFC 8785 JSON Canonicalization Scheme — self contained implementation.

Independent of the browser implementation on purpose. Cross language parity is
only meaningful if the two sides were written separately from the spec.
"""
from __future__ import annotations

import math

# RFC 8785 §3.2.2.2 — the two-character escapes, everything else below 0x20
# becomes \u00xx lowercase hex.
_SHORT_ESCAPES = {
    0x08: "\\b",
    0x09: "\\t",
    0x0A: "\\n",
    0x0C: "\\f",
    0x0D: "\\r",
    0x22: '\\"',
    0x5C: "\\\\",
}


def _escape(s: str) -> str:
    out = ['"']
    for ch in s:
        cp = ord(ch)
        if cp in _SHORT_ESCAPES:
            out.append(_SHORT_ESCAPES[cp])
        elif cp < 0x20:
            out.append("\\u%04x" % cp)
        else:
            out.append(ch)
    out.append('"')
    return "".join(out)


def _number(n) -> str:
    """ECMAScript Number::toString. Money is integer paise, so the integer path
    is the one that carries load; the float path is here so the function is
    honest rather than because we intend to use it."""
    if isinstance(n, bool):
        raise TypeError("bool is not a JSON number")
    if isinstance(n, int):
        if abs(n) > 2 ** 53 - 1:
            raise ValueError("integer outside IEEE-754 exact range: %d" % n)
        return str(n)
    if not math.isfinite(n):
        raise ValueError("NaN and Infinity are not valid JSON")
    if n == int(n) and abs(n) < 1e21:
        return str(int(n))
    r = repr(float(n))
    if "e" in r:  # python 1e+21 / 1e-07  ->  ES6 1e+21 / 1e-7
        mant, exp = r.split("e")
        sign = "+" if not exp.startswith("-") else "-"
        r = "%se%s%d" % (mant, sign, abs(int(exp)))
    return r


def _utf16_key(s: str) -> bytes:
    """RFC 8785 sorts object keys by UTF-16 code unit, which is exactly
    lexicographic order over the big endian UTF-16 encoding."""
    return s.encode("utf-16-be", errors="surrogatepass")


def _ser(v, out: list) -> None:
    if v is None:
        out.append("null")
    elif v is True:
        out.append("true")
    elif v is False:
        out.append("false")
    elif isinstance(v, str):
        out.append(_escape(v))
    elif isinstance(v, (int, float)):
        out.append(_number(v))
    elif isinstance(v, (list, tuple)):
        out.append("[")
        for i, item in enumerate(v):
            if i:
                out.append(",")
            _ser(item, out)
        out.append("]")
    elif isinstance(v, dict):
        out.append("{")
        for i, k in enumerate(sorted(v.keys(), key=_utf16_key)):
            if not isinstance(k, str):
                raise TypeError("object keys must be strings")
            if i:
                out.append(",")
            out.append(_escape(k))
            out.append(":")
            _ser(v[k], out)
        out.append("}")
    else:
        raise TypeError("not JSON serialisable: %r" % type(v))


def canonicalize(value) -> bytes:
    out: list = []
    _ser(value, out)
    return "".join(out).encode("utf-8")

File: console/src/lib/jcs.ts
/**
 * RFC 8785 JSON Canonicalization Scheme.
 *
 * Written from the RFC, not ported from the Python side, because parity only
 * proves something if the two implementations are independent.
 *
 * Two traps this avoids, both called out in LANE-C section 1:
 *   - JSON.stringify with sorted keys is NOT JCS. It sorts by whatever the
 *     caller's comparator does and it does not normalise numbers.
 *   - Key ordering is by UTF-16 code unit. JavaScript's default string sort
 *     already compares UTF-16 code units, which is the one place the platform
 *     happens to agree with the RFC. Do not "improve" it with localeCompare.
 */

export type JsonValue =
  | null
  | boolean
  | number
  | string
  | JsonValue[]
  | { [k: string]: JsonValue };

// RFC 8785 §3.2.2.2
const SHORT_ESCAPES: Record<number, string> = {
  0x08: "\\b",
  0x09: "\\t",
  0x0a: "\\n",
  0x0c: "\\f",
  0x0d: "\\r",
  0x22: '\\"',
  0x5c: "\\\\",
};

function escapeString(s: string): string {
  let out = '"';
  for (const ch of s) {
    const cp = ch.codePointAt(0)!;
    const short = SHORT_ESCAPES[cp];
    if (short !== undefined) out += short;
    else if (cp < 0x20) out += "\\u" + cp.toString(16).padStart(4, "0");
    else out += ch;
  }
  return out + '"';
}

function serializeNumber(n: number): string {
  if (!Number.isFinite(n)) throw new Error("NaN and Infinity are not valid JSON");
  // RFC 8785 §3.2.2.3: ES6 Number::toString, with -0 normalised to 0.
  if (n === 0) return "0";
  return String(n);
}

function ser(v: JsonValue, out: string[]): void {
  if (v === null) {
    out.push("null");
  } else if (typeof v === "boolean") {
    out.push(v ? "true" : "false");
  } else if (typeof v === "number") {
    out.push(serializeNumber(v));
  } else if (typeof v === "string") {
    out.push(escapeString(v));
  } else if (Array.isArray(v)) {
    out.push("[");
    v.forEach((item, i) => {
      if (i) out.push(",");
      ser(item, out);
    });
    out.push("]");
  } else if (typeof v === "object") {
    out.push("{");
    // Default sort compares UTF-16 code units, which is what the RFC asks for.
    const keys = Object.keys(v).sort();
    keys.forEach((k, i) => {
      if (i) out.push(",");
      out.push(escapeString(k));
      out.push(":");
      ser(v[k], out);
    });
    out.push("}");
  } else {
    throw new Error(`not JSON serialisable: ${typeof v}`);
  }
}

/** Canonical JSON as a string. */
export function canonicalizeToString(value: JsonValue): string {
  const out: string[] = [];
  ser(value, out);
  return out.join("");
}

/** Canonical JSON as UTF-8 bytes. This is what gets signed. */
export function canonicalize(value: JsonValue): Uint8Array {
  return new TextEncoder().encode(canonicalizeToString(value));
}
